### ISSUE TITLE
[RFC] API for canceling HTTP requests

### ISSUE THREAD

**Issue by nickuraltsev** [mzabriskie/axios#333]
# Overview

I would like to revive the discussion about canceling HTTP requests (#50) as there is clearly a demand for this feature. I suggest considering the following options. /cc @mzabriskie 

### API

```javascript
var axios = require('axios');

function generateRequestId() {
  // Return a unique request ID
}

var requestId = generateRequestId(); 
axios.get('/user', { requestId: requestId });
axios.abort(requestId);
```

### Implementation

PR #330

### API

```javascript
var axios = require('axios');
var Cancellation = axios.Cancellation;

var cancellation = new Cancellation();
axios.get('/user', { cancellation: cancellation });
cancellation.cancel();
```

### Notes
- When a request is canceled, the `Promise` is rejected with a `CancellationError`. This allows the caller to distinguish between a cancellation and a failure.
- The same `Cancellation` object can be used to cancel multiple different requests. 

### Implementation

I created a separate library that can be used as a base for implementing request cancellation in axios: https://github.com/nickuraltsev/cancel. (The [example](https://github.com/nickuraltsev/cancel#example) on the README.md shows how to implement a simple HTTP client that supports cancellation.) We can copy all the code from this library to axios or add this library as a dependency.

**Comment by mzabriskie** [comment]
## Option 3

### API

```js
import axios from 'axios'

let cancelRequest
let promise = new Promise((resolve) => {
  cancelRequest = resolve
});

window.addEventListener('beforeunload', (e) => {
  cancelRequest()
})

axios.get('/user', {
  timeout: promise
})
```

### Notes

- If the `Promise` is resolved the request is canceled. This is what Angular does with `$http` service.

### Implementation

The implementation for this is simply checking `config.timeout instanceof Promise` and if it is adding a `then` to it which if called cancels the request that is being made.

**Comment by msangui** [comment]
Abort on retry can be even more important than abort by itself. For example the most common use case for aborting XHR requests is when you are doing an autocomplete logic, we should discuss that feature as well.

**Comment by mzabriskie** [comment]
I just saw that TC39 is discussing a proposal for [Cancelable Promises](https://docs.google.com/presentation/d/1V4vmC54gJkwAss1nfEt9ywc-QOVOfleRxD5qtpMpc8U/edit?usp=sharing)

**Comment by nickuraltsev** [comment]
@mzabriskie Yes, I'm aware of that proposal. It consists of two parts (slide 23):
- Cancellation as a third state 
- Cancellation tokens

I don't think we can use cancelable promises in axios API before they are implemented in the browsers (I may be wrong). But I do think that we can use cancellation tokens to request cancellation. That is what Option 2 is about. I think we should align the cancellation token API to match the one described in the proposal though. 

As for the third state, I would suggest to reject promises with a special error (e.g. CancellationError) for now. Once cancellable tokens are implemented, we can update axios to cancel the promise instead of rejecting it. This is not going to happen soon, so I don't think we should wait for it.

Thoughts?

**Comment by diessica** [comment]
**Option 1** seems good enough for me. Also liked [@herrstucki's spproach](https://gist.github.com/mzabriskie/ec3a25dd45dfa0de92c2#gistcomment-1416322) to the issue.

**Comment by pasin** [comment]
Is there a timeline for the abort API to be finalized and implemented? IMO, it is a good idea to have the API inline with the Cancelable Promise spec.

**Comment by just-boris** [comment]
I think it is possible to implement any option as a custom adapter (based on XHR) and then try it in a real project.

**Comment by asd0102433** [comment]
i think Option 1 is good idea . @mzabriskie

**Comment by just-boris** [comment]
I tried to use `Option 2` in my project and found it nice. If anybody else would like to try this out, I published this as project [just-boris/axios-cancel](https://github.com/just-boris/axios-cancel). It is custom adapter which accepts cancellation object in options. See installation instructions in [readme](https://github.com/just-boris/axios-cancel/blob/master/README.md) to get started

**Comment by beckiechoi** [comment]
I also second Option 2, it's more future proof.

**Comment by jeffijoe** [comment]
I like the idea of cancellation tokens as @just-boris is doing - this is also how it's done in .NET.

Basically make a cancellation token and pass it in, and when `cancel()` is called, abort the request and reject with a `CancellationError`.

**Comment by andrewmclagan** [comment]
Option 2 is really  awesome...!

**Comment by just-boris** [comment]
I think that while promise calcellation spec is not finalized, this feature will not be implemented in core. To gracefully adopt cancellation feature, it would be better to use this as plugin, like I did.

If anybody interested, I can publish [axios-cancel](https://github.com/just-boris/axios-cancel) on NPM for convinience.

**Comment by nickuraltsev** [comment]
I believe this feature should be implemented in core. There is no way to build it as a plugin without re-implementing axios adapters as it's done in axios-cancel which seems to be more a fork rather than a plugin. 

Here is my suggestion:

- Short-term plan: 
  - Add support for cancellation right now
  - Implementation should be as close as possible to [Cancelable promises proposal](https://github.com/domenic/cancelable-promise)
- Long-term plan: 
  - Once cancellation is standardized and implemented in browsers/node, we can update axios to use built-in cancellation primitives. (It will be a breaking change so we will need to bump the version number appropriately.)

I think that the short-term solution doesn't need to be perfect. The goal is to make it possible to cancel requests and handle cancellations. Once it's in place, users will have a choice of using the axios API as is or build another API on top of it. Now, there are no options except re-implementing the adapter as @just-boris did. 

@mzabriskie If you agree with this approach, I can create a PR. Please let me know what you think.

**Comment by mzabriskie** [comment]
I don't think that creating offshoot adapter implementations is the correct approach. This will be difficult to maintain and keep synchronized with the core adapters. There will also need to be an offshoot maintained for both node's http as well as browser's xhr.

I agree with having a short-term and long-term plan. I support @nickuraltsev's proposal above.

**Comment by nickuraltsev** [comment]
Great, I'll submit a PR soon.

**Comment by miraage** [comment]
@nickuraltsev what's current progress status?

**Comment by nhducit** [comment]
I vote for option 1 or option 2

**Comment by pavel06081991** [comment]
Any news?

**Comment by nickuraltsev** [comment]
Just a heads up that I'm actively working on this feature. I'm hoping it will be ready in a couple of days.

**Comment by nickuraltsev** [comment]
I've just submitted a PR

**Comment by wlingke** [comment]
Any news on when this will be released in an official version?

**Comment by nickuraltsev** [comment]
Fixed via #452

**Comment by comerc** [comment]
[FYI](https://medium.com/@benlesh/promise-cancellation-is-dead-long-live-promise-cancellation-c6601f1f5082)
